### ISSUE TITLE
Channel graph issue for 40Mhz

### ISSUE THREAD

**Issue by JJWatMyself** [VREMSoftwareDevelopment/WiFiAnalyzer#58]
Channel graph issues:

I noticed that pre 1.6.2 that a 2.4Ghz 40MHz SSID was displaying -1 channel of the actual and now on 1.6.2 it is -2 channels of the actual.

Thanks to https://github.com/arafey for pointing out that the extension channel is four channels lower or higher than the control channel.

[arafey](https://github.com/arafey) **"A 40 MHz network wouldn't use 11+9. It would use 11+7. 9 is the center of the two channels. Some wireless APs will show 11+9 but they're telling you the primary channel and the center channel."**
![understanding-rf-fundamentals-and-the-radio-design-of-wireless-networks-60-728](https://cloud.githubusercontent.com/assets/7505951/19043185/1ca5d3be-895d-11e6-8627-6f1fc04277a4.jpg)

**Comment by arafey** [comment]
You're seeing a 40 MHz AP that has a primary channel of 11 show up on the graph as channel 9 (see the screenshot below)?

![screenshot_20161003-134756](https://cloud.githubusercontent.com/assets/7505951/19047775/3d540be4-8971-11e6-9626-214a2e60d86b.png)

If so, then that's supposed to be the case because channel 9 is the center of the two channels (11+7) and forms the wide channel. The screenshot also shows another AP that is on channel 1 with an extension of 5. It's offset by +2 because channel 3 is the center there. I opened an issue recently (#56, which was resolved) about proper display of 40+ MHz networks.

If what you're seeing is different, then a screenshot would be helpful.

**Comment by VREMSoftwareDevelopment** [comment]
Thank you [arafey](https://github.com/arafey),

The VREM Development Team

**Comment by JJWatMyself** [comment]
Your config matches mine.

AP1 = control chan 1, extension chan 5, center 3
AP2 = control chan 11, extension chan 7, center 9

And interestingly your screen shot is correct.

I have dd-wrt on two asus ac68u. Wifi Analyzer does not draw them correctly though.  Notice how the SSID shows the correct channel number above them though.
![screenshot_20161003-123347](https://cloud.githubusercontent.com/assets/22590445/19056835/97aac538-8999-11e6-9fe4-3dc048a9863e.png)
![screenshot_20161003-184513](https://cloud.githubusercontent.com/assets/22590445/19056836/97ab2bc2-8999-11e6-8c24-ea48fbf3874f.png)

**Comment by JJWatMyself** [comment]
Maybe this is a device, or OS specific issue as I notice the same issue in another well known Android wifi analyzer app.

Samsung Galaxy S6
Marshmallow 6.0.1

**Comment by JJWatMyself** [comment]
![tmp_8441-screenshot_20161003-202803-430469992](https://cloud.githubusercontent.com/assets/22590445/19058754/ef384484-89a7-11e6-9fbc-66104a530bf0.png)

**Comment by JJWatMyself** [comment]
In the other app if I turn off the feature to show channel width the correct control channel is displayed.
![tmp_8441-screenshot_20161003-203436429719306](https://cloud.githubusercontent.com/assets/22590445/19058878/1cc80302-89a9-11e6-9fef-1810767eb40e.png)

**Comment by VREMSoftwareDevelopment** [comment]
[Developer Android API](https://developer.android.com/reference/android/net/wifi/ScanResult.html) for OS 6 and higher provides the following:
- Primary frequency: the primary frequency of the channel over which the client is communicating with the access point. 
- Center frequency: the center frequency used if the AP bandwidth is 40, 80 or 160 MHz. 

The WiFi Analyzer uses the values provided by API to display information in the Access Points and Channel Graph.

Most unlikely it is a device, or OS specific issue. Most likely it is access point software (DD-WRT, OpenWrt, Tomato, etc) issue: Software does not set primary and center frequency/channel correctly or Software does not transmit primary and center frequency/channel correctly. Please note this is only speculations.

**Comment by JJWatMyself** [comment]
Thanks for the feedback.  Did you notice the screenshot I posted display the correct control channel?

**Comment by VREMSoftwareDevelopment** [comment]
1. Primary/Control channel is equals to center channel when using 20MHz channel wide.
2. Primary/Control channel does not necessarily equals to center channel when using 40/80/160 MHz channel wide. That exactly what you are seeing.

**Comment by arafey** [comment]
I am running DD-WRT Kong builds and I used to run Tomato Shibby builds, Advanced Tomato, Tomato stock, and DD-WRT BS/Eko builds as well. Our setups should be comparable. 

To rule out the AP as an issue, can you confirm that your DD-WRT Wireless Setup and the Sys-Info page have similar configs to mine?

*Sys-Info*
![screenshot_20161005-203925 2](https://cloud.githubusercontent.com/assets/7505951/19136928/3a4c3416-8b3e-11e6-8807-792f4cfd0091.png)
![screenshot_20161005-203821 2](https://cloud.githubusercontent.com/assets/7505951/19136927/3a4c14ae-8b3e-11e6-907a-870d218e3f06.png)
![screenshot_20161005-203837 2]

*Wireless Setup*
(https://cloud.githubusercontent.com/assets/7505951/19136929/3a4cf7fc-8b3e-11e6-9583-c53d4c45eab1.png)
![screenshot_20161005-203910 2](https://cloud.githubusercontent.com/assets/7505951/19136930/3a4f8b84-8b3e-11e6-9919-631206905e84.png)

What version of DD-WRT are you running currently? Mine are all on Kong r30700m. Your router may only have BrainSlayer builds, which may have a different version number. If it's recent, it should be >30000. 

I know that the AC68U has good open source firmware compatibility and tends to receive lots of support from the projects. I think it's not very likely that a properly configured router of this model running a recent and relatively stable build of DD-WRT (check their forums under Broadcom-based Hardware for people's experience with various builds) would have a massive issue such as incorrect channel selection or running a non-802.11 compliant network (which it would be if the primary and secondary channels are more than 4 apart on 2.4 GHz radios). In my opinion, in that case, it would warrant further investigation into the device or the WiFi Analyzer app.

Tl;dr: Compare the screenshots with your AP. Check your firmware version and look at the DD-WRT forum for a good recent version or any hint of 40 MHz/HT, channel selection/display, etc. problems with your particular router. If all is well, then you should report back here.

**Comment by JJWatMyself** [comment]
Thanks for the feedback Arafey and apologies for the delay in my response.  My config is identical to yours.  I did some additional tests this morning.  When I originally posted, the two AC86U that I have were running BrainSlayer 29519.  I upgraded to 30771 and observed the same results.  Then I factory reset (erase nvram) and change to Kong v3.0-r30700M kongac (09/27/16 dd-wrt.v24-K3_AC_ARM_STD.bin).  Same results.

**Comment by VREMSoftwareDevelopment** [comment]
@JJWatMyself  Could you please post your DD-WRT Wireless Setup screenshots from the router (same as @arafey) ?

**Comment by JJWatMyself** [comment]
![net-1-1](https://cloud.githubusercontent.com/assets/22590445/19621484/0e9fc910-9861-11e6-8a09-d0f4a43e5c05.PNG)
![net-1-2](https://cloud.githubusercontent.com/assets/22590445/19621485/100fcb24-9861-11e6-8603-b692efb2f4ce.PNG)

**Comment by VREMSoftwareDevelopment** [comment]
We are seeing the same issues on some of the routers, but not all of them, the router UI shows 11+9, Android API returns 11+7.

**Comment by JJWatMyself** [comment]
```
root@hamster-net-1:/# wl -i eth1 status
SSID: "hamster-n"
Mode: Managed   RSSI: 0 dBm     SNR: 0 dB       noise: -89 dBm  Channel: 11u
BSSID: 78:24:AF:7A:16:2A        Capability: ESS ShortSlot RRM
Supported Rates: [ 1(b) 2(b) 5.5(b) 6(b) 9 11(b) 12(b) 18 24(b) 36 48 54 ]
HT Capable:
        Chanspec: 2.4GHz channel 9 40MHz (0x1909)
        Primary channel: 11
        HT Capabilities:
        Supported MCS : [ 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 32 ]
```

**Comment by JJWatMyself** [comment]
Here is my other AC68U which provides coverage area on the other side of the house.

```
root@hamster-net-2:~# wl -i eth1 status
SSID: "hamster-n"
Mode: Managed   RSSI: 0 dBm     SNR: 0 dB       noise: -91 dBm  Channel: 1l
BSSID: 34:97:F6:60:D9:82        Capability: ESS ShortSlot RRM
Supported Rates: [ 1(b) 2(b) 5.5(b) 6(b) 9 11(b) 12(b) 18 24(b) 36 48 54 ]
HT Capable:
        Chanspec: 2.4GHz channel 3 40MHz (0x1803)
        Primary channel: 1
        HT Capabilities:
        Supported MCS : [ 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 32 ]

```

**Comment by VREMSoftwareDevelopment** [comment]
Looks like Android API when using 2.4GHz does not return center channel in their API, they are returning Ext Channel, but not for all the routers.

**Comment by JJWatMyself** [comment]
Do you think that this is router specific or client specific?  Thanks.  :)

**Comment by arafey** [comment]
Just to clarify, is `centerFreq0` returning the extension channel in some scenarios?
I think a conditional could help:
```
let centerFreq = centerFreq0;
if centerFreq0 == frequency + 20 {
    centerFreq -= 10
}
else if centerFreq0 == frequency - 20 {
    centerFreq += 10
}
```

Beyond this point, all devices will have the same `centerFreq` values and should output the same information in a 2.4 GHz environment with 40 MHz channels.

Also @JJWatMyself, I would be inclined to say it's client-specific, but if @VREMSoftwareDevelopment has the same client showing 11+7 in one circumstance and 11+9 in another, then it could be a combination of  buggy clients along with specific APs.

**Comment by VREMSoftwareDevelopment** [comment]
Probably Android API, since in the documentation it states center frequency, but it is actually returns ext frequency.
